### PR TITLE
[Scheduler] New regression tests (particularly around update and pause)

### PR DESCRIPTION
**Pull request by lina-temporal**
### Functional tests (`tests/schedule_test.go`)
The full edge-case suite for the reworked behaviors, run against V1 and CHASM where applicable: trigger-immediately on paused/exhausted schedules, backfill variants (SKIP, BUFFER_ALL, narrow-range, multi-range counted-exactly-once), schedule-closes-from-idle (4 spec shapes), pause/update-during-idle, manual-only close, backfill-blocks-idle, recent-actions / future-action-times while paused. 

Also parallelizes the schedule suite (367s → 67s).

### Unit tests (`chasm/lib/scheduler/`)
Additional coverage of reworked internals: the ExecuteTask/ProcessBuffer `Validate` tables, paused-drops-automated, generator-paused-advances-HWM, the new idle held-open / sentinel cases, and the Nexus-completion deferred-re-enable cascade.

### CODE REVIEW

**Comment by chaptersix** [comment]
@lina-temporal could you rebase?

File: chasm/lib/scheduler/generator_tasks_test.go [hunk 1]
```diff
@@ -197,6 +197,38 @@ func TestGeneratorTask_UpdateFutureActionTimes_SkipsBeforeUpdateTime(t *testing.
 	}
 }
 
+// TestGeneratorTask_PausedDropsActionsAdvancesHWM verifies the "drop, don't
+// buffer; keep advancing the HWM" semantic of the always-on GeneratorTask.
+// A paused Execute over a non-trivial time range must produce no buffered
+// starts and must still advance the high water mark.
```
Comment (davidporter-id-au) [inline chasm/lib/scheduler/generator_tasks_test.go:203]: Question for my learning: high-water-mark is `LastProcessedTime`? And I assume this implies the invariant: The high-water-mark never decreases?

File: chasm/lib/scheduler/invoker_process_buffer_task_test.go [hunk 2]
```diff
@@ -17,6 +17,37 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+// Validate is currently the only line of defense against a stale ProcessBuffer
```
Comment (davidporter-id-au) [inline chasm/lib/scheduler/invoker_process_buffer_task_test.go]: Might need a little more exposition as to what this dire-sounding warning is referring to. 

Last line of defense of incorrectly processing a task twice?

File: tests/schedule_test.go [hunk 3]
```diff
@@ -202,6 +221,205 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+	sid := testcore.RandomizeStr("sched-recentactions-paused")
+	wid := testcore.RandomizeStr("sched-recentactions-paused-wf")
+	wt := testcore.RandomizeStr("sched-recentactions-paused-wt")
+
+	s.SdkWorker().RegisterWorkflowWithOptions(
+		func(ctx workflow.Context) error {
+			return workflow.Sleep(ctx, 3*time.Second)
+		},
+		workflow.RegisterOptions{Name: wt},
+	)
```
Comment (chaptersix) [inline tests/schedule_test.go]: could you use a signal instead of a sleep. ideally this is time independent.

File: tests/schedule_test.go [hunk 4]
```diff
@@ -202,6 +221,205 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+
+	ctx := newContext(s.Context())
+	_, err := s.FrontendClient().CreateSchedule(ctx, &workflowservice.CreateScheduleRequest{
+		Namespace:  s.Namespace().String(),
+		ScheduleId: sid,
+		Schedule: &schedulepb.Schedule{
+			Spec: &schedulepb.ScheduleSpec{
+				Interval: []*schedulepb.IntervalSpec{{Interval: durationpb.New(1 * time.Second)}},
```
Comment (chaptersix) [inline tests/schedule_test.go]: could we store this spec as a const? `fastSpec` or something similar?

File: tests/schedule_test.go [hunk 5]
```diff
@@ -202,6 +221,205 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+	})
+	require.NoError(t, err)
+
+	// Wait for the first workflow to start and a second to be buffered behind it.
+	s.Eventually(func() bool {
+		desc, descErr := s.FrontendClient().DescribeSchedule(ctx, &workflowservice.DescribeScheduleRequest{
+			Namespace:  s.Namespace().String(),
+			ScheduleId: sid,
+		})
+		return descErr == nil && desc.GetInfo().GetBufferSize() >= 1 && len(desc.GetInfo().GetRunningWorkflows()) >= 1
```
Comment (chaptersix) [inline tests/schedule_test.go]: can we be exact here?

File: tests/schedule_test.go [hunk 6]
```diff
@@ -202,6 +221,205 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+				OverlapPolicy: enumspb.SCHEDULE_OVERLAP_POLICY_BUFFER_ONE,
+			},
+		},
+		Identity:  "test",
+		RequestId: uuid.NewString(),
+	})
+	require.NoError(t, err)
+
+	// Wait for the first workflow to start and a second to be buffered behind it.
+	s.Eventually(func() bool {
```
Comment (chaptersix) [inline tests/schedule_test.go]: not sure why the lint's are not firing for this but should you use the `await` instead of `Eventually`

File: tests/schedule_test.go [hunk 7]
```diff
@@ -202,6 +221,205 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+
+	sid := testcore.RandomizeStr("sched-buffer-one-deferred")
+	wid := testcore.RandomizeStr("sched-buffer-one-deferred-wf")
+	wt := testcore.RandomizeStr("sched-buffer-one-deferred-wt")
+
+	var runs atomic.Int32
+	s.SdkWorker().RegisterWorkflowWithOptions(
+		func(ctx workflow.Context) error {
+			runs.Add(1)
+			return workflow.Sleep(ctx, 3*time.Second)
```
Comment (chaptersix) [inline tests/schedule_test.go]: this also seems like one we can make non-time dependent.

File: tests/schedule_test.go [hunk 8]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+	// workflow before close, whose runs.Add() may not have landed yet at the
+	// moment DescribeSchedule first errored.
+	s.Never(func() bool { return runs.Load() > 0 },
+		2*time.Second, 100*time.Millisecond,
+		"paused-window calendar entry must not be replayed on unpause")
+}
+
+// testPausedScheduleNeverIdles verifies that a paused schedule is held open
+// indefinitely, even past the configured IdleTime. The CHASM tweakable is set
+// here unconditionally; on V1 it has no effect (V1 schedules don't auto-close),
```
Comment (chaptersix) [inline tests/schedule_test.go]: > V1 schedules don't auto-close

not sure I follow this part. they self close after idle time has expired.

File: tests/schedule_test.go [hunk 9]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+	wid := testcore.RandomizeStr("sched-paused-never-idles-wf")
+	wt := testcore.RandomizeStr("sched-paused-never-idles-wt")
+
+	var runs atomic.Int32
+	s.SdkWorker().RegisterWorkflowWithOptions(
+		func(ctx workflow.Context) error {
+			workflow.SideEffect(ctx, func(ctx workflow.Context) any {
+				runs.Add(1)
+				return 0
+			})
```
Comment (chaptersix) [inline tests/schedule_test.go]: why do some tests inc the count in a side effect and some do not?

File: tests/schedule_test.go [hunk 10]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+		},
+		Identity:  "test",
+		RequestId: uuid.NewString(),
+	})
+	require.NoError(t, err)
+
+	// Confirm the paused schedule isn't firing on its own.
+	s.Never(func() bool { return runs.Load() > 0 },
+		2*time.Second, 100*time.Millisecond,
+		"paused schedule must not fire automated actions before the trigger")
```
Comment (chaptersix) [inline tests/schedule_test.go]: we need a shorter (1s) interval if we want to assert this.

File: tests/schedule_test.go [hunk 11]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+		workflow.RegisterOptions{Name: wt},
+	)
+
+	ctx := newContext(s.Context())
+	_, err := s.FrontendClient().CreateSchedule(ctx, &workflowservice.CreateScheduleRequest{
+		Namespace:  s.Namespace().String(),
+		ScheduleId: sid,
+		Schedule: &schedulepb.Schedule{
+			Spec: &schedulepb.ScheduleSpec{
+				Interval: []*schedulepb.IntervalSpec{{Interval: durationpb.New(10 * time.Minute)}},
```
Comment (chaptersix) [inline tests/schedule_test.go]: const for this interval: `noOpInterval` or something similar?

File: tests/schedule_test.go [hunk 12]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+			},
+		},
+		Identity:  "test",
+		RequestId: uuid.NewString(),
+	})
+	require.NoError(t, err)
+
+	s.Never(func() bool { return runs.Load() > 0 },
+		2*time.Second, 100*time.Millisecond,
+		"exhausted schedule must not auto-fire before the trigger")
```
Comment (chaptersix) [inline tests/schedule_test.go]: need a 1s interval to test this.

File: tests/schedule_test.go [hunk 13]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+		workflow.RegisterOptions{Name: wt},
+	)
+
+	ctx := newContext(s.Context())
+	_, err := s.FrontendClient().CreateSchedule(ctx, &workflowservice.CreateScheduleRequest{
+		Namespace:  s.Namespace().String(),
+		ScheduleId: sid,
+		Schedule: &schedulepb.Schedule{
+			Spec: &schedulepb.ScheduleSpec{
+				// 1-hour interval; the backfill range below is only 100ms wide.
```
Comment (chaptersix) [inline tests/schedule_test.go]: 100ms is lower than 1s / our highest resolution interval. Seems like a different test than what the descriptions says.

File: tests/schedule_test.go [hunk 14]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+				OverlapPolicy: enumspb.SCHEDULE_OVERLAP_POLICY_SKIP,
+			}},
+		},
+		Identity:  "test",
+		RequestId: uuid.NewString(),
+	})
+	require.NoError(t, err)
+
+	// SKIP collapses the 5-tick backfill to a single fire.
+	s.Eventually(func() bool { return runs.Load() >= 1 },
```
Comment (chaptersix) [inline tests/schedule_test.go]:
```suggestion
	s.Eventually(func() bool { return runs.Load() == 1 },
```

File: tests/schedule_test.go [hunk 15]
```diff
@@ -3479,41 +3701,792 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+			},
+			State: &schedulepb.ScheduleState{Paused: true},
+			Action: &schedulepb.ScheduleAction{
+				Action: &schedulepb.ScheduleAction_StartWorkflow{
+					StartWorkflow: &workflowpb.NewWorkflowExecutionInfo{
+						WorkflowId:            wid,
+						WorkflowType:          &commonpb.WorkflowType{Name: wt},
+						TaskQueue:             &taskqueuepb.TaskQueue{Name: s.WorkerTaskQueue(), Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+						WorkflowIdReusePolicy: enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+					},
```
Comment (chaptersix) [inline tests/schedule_test.go]: should we have func or set of funcs that produce a default action? Might be easier to identify what setting is specific to a given test.

File: tests/schedule_test.go [hunk 16]
```diff
@@ -3674,193 +4647,509 @@ func TestScheduleCreationRolloutPercent(t *testing.T) {
...
+				},
+			},
+		},
+		Identity:  "test",
+		RequestId: uuid.NewString(),
+	})
+	require.NoError(t, err)
+
+	s.Eventually(func() bool { return runs.Load() >= c.expectedRuns },
```
Comment (chaptersix) [inline tests/schedule_test.go]:
```suggestion
	s.Eventually(func() bool { return runs.Load() == c.expectedRuns },
```

File: tests/schedule_test.go [hunk 17]
```diff
@@ -3674,193 +4647,509 @@ func TestScheduleCreationRolloutPercent(t *testing.T) {
...
+		s.Equal(c.expectedRuns, runs.Load(), "schedule must not exceed its action budget")
+	}
+}
+
+// testManualOnlyUnpausedClosesFromIdle verifies that an unpaused manual-only
+// (empty-spec) schedule closes after its idle/retention window elapses on both
+// V1 and CHASM. V1 advances lastEvent via RecentActions, V1 RetentionTime is
+// plumbed via WorkerSchedulerRetentionTime; CHASM uses IdleTime. The two
+// values are set to the same short window so the test is symmetric.
+func testManualOnlyUnpausedClosesFromIdle(t *testing.T, newContext contextFactory) {
```
Comment (chaptersix) [inline tests/schedule_test.go:4521]: this seems to conflict with a functional tests earlier. Should an empty spec schedule say open or close after idle?

File: tests/schedule_test.go [hunk 18]
```diff
@@ -3674,193 +4647,509 @@ func TestScheduleCreationRolloutPercent(t *testing.T) {
...
+		})
+		return err == nil && len(resp.Info.FutureActionTimes) == 0
+	}, 15*time.Second, 200*time.Millisecond, "schedule should drain its future action times")
+
+	s.Eventually(func() bool {
+		_, err := s.FrontendClient().DescribeSchedule(ctx, &workflowservice.DescribeScheduleRequest{
+			Namespace:  s.Namespace().String(),
+			ScheduleId: sid,
+		})
+		return err != nil
```
Comment (chaptersix) [inline tests/schedule_test.go]: explicit error check?

File: tests/schedule_test.go [hunk 19]
```diff
@@ -3674,193 +4647,509 @@ func TestScheduleCreationRolloutPercent(t *testing.T) {
...
+		RequestId: uuid.NewString(),
+	})
+	require.NoError(t, err)
+
+	s.Eventually(func() bool {
+		_, err := s.FrontendClient().DescribeSchedule(ctx, &workflowservice.DescribeScheduleRequest{
+			Namespace:  s.Namespace().String(),
+			ScheduleId: sid,
+		})
+		return err != nil
```
Comment (chaptersix) [inline tests/schedule_test.go]: explicit not found check?

File: tests/schedule_test.go [hunk 20]
```diff
@@ -3674,193 +4647,509 @@ func TestScheduleCreationRolloutPercent(t *testing.T) {
...
+}
+
+// testPauseDuringIdleWindow exercises the held_open invalidation + re-arm path.
+// A LimitedActions=1 schedule fires once; the Generator arms an idle task.
+// Before the idle deadline elapses, we pause: any in-flight idle task that
+// fires must be dropped by Validate with reason=held_open. After unpause, the
+// Generator re-arms and the schedule eventually closes. IdleTime is sized
+// generously so the Pause RPC arrives well before the original idle deadline
+// even under slow CI.
+func testPauseDuringIdleWindow(t *testing.T, newContext contextFactory) {
```
Comment (chaptersix) [inline tests/schedule_test.go]: this is a good example of a test that it would be great to have time skip support for. I'm think of the chasm test engine in particular.

**Review by chaptersix — COMMENTED** [review]
I've reviewed just the functional tests so fair. Quite a few comments to respond to and getting a bit fatigued so I'll stop here for now. 

I'll review non functional tests later today or tomorrow.

**Comment by lina-temporal** [comment]
> I've reviewed just the functional tests so fair. Quite a few comments to respond to and getting a bit fatigued so I'll stop here for now.
> 
> I'll review non functional tests later today or tomorrow.

Ah, sorry about that - feel free to hold off for now, I'm going to take another pass myself first. Truth be told, I was pretty fatigued while ghostwriting them.

File: tests/schedule_test.go [hunk 21]
```diff
@@ -3479,107 +3803,486 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+// testBackfillWithBufferOneOverlap pins the expected behavior of BUFFER_ONE
+// over a multi-tick backfill: the first start runs immediately, exactly one
+// follow-up is buffered (Attempt=-1 deferred), the rest are dropped, and the
+// deferred one runs once the first completes. Currently SKIPPED: fails on
+// both V1 and CHASM because the deferred start never gets re-enabled after
+// the running workflow completes. The first start fires, the rest never run.
+// Likely a real bug in the BUFFER_ONE + backfill (Manual=true) interaction -
+// recordCompletedAction's re-enable loop on Attempt==-1 may not be running
+// against backfill-buffered starts. Worth a separate investigation.
+func testBackfillWithBufferOneOverlap(t *testing.T, newContext contextFactory) {
```
Comment (lina-temporal) [inline tests/schedule_test.go:4109]: Planning to fast-follow with this, added a task for it

File: chasm/lib/scheduler/invoker_execute_task_test.go [hunk 22]
```diff
@@ -528,6 +528,87 @@ func TestExecuteTask_Validate_BackoffEqualToLPTIsEligible(t *testing.T) {
...
+				invoker.LastProcessedTime = timestamppb.New(now)
+				invoker.BufferedStarts = []*schedulespb.BufferedStart{{
+					RequestId: "ready",
+					Attempt:   1,
+				}}
+			},
+			expectedValid: true,
+		},
+		{
+			name: "running start (RunId set) is invalid",
```
Comment (davidporter-id-au) [inline chasm/lib/scheduler/invoker_execute_task_test.go:562]: Why is this not valid? It's probably very simple, but I don't understand

File: chasm/lib/scheduler/invoker_execute_task_test.go [hunk 23]
```diff
@@ -528,6 +528,87 @@ func TestExecuteTask_Validate_BackoffEqualToLPTIsEligible(t *testing.T) {
...
+				invoker.BufferedStarts = []*schedulespb.BufferedStart{{
+					RequestId: "running",
+					Attempt:   1,
+					RunId:     "run-1",
+				}}
+			},
+			expectedValid: false,
+		},
+		{
+			name: "backing-off start is invalid",
```
Comment (davidporter-id-au) [inline chasm/lib/scheduler/invoker_execute_task_test.go:575]: It's also not obvious to me why this is invalid, would you be able to update the context string as to the reason?

After poking around a bit / getting help from Claude I assume it's that this is reason this is considered invalid is that the backoff time is 1m, but the last time this was processed was now, so it's too early, is that correct and it's safe for us to drop this task owing to these tasks being generated by the Invoker task?

File: chasm/lib/scheduler/scheduler_nexus_completion_test.go [hunk 24]
```diff
@@ -253,6 +253,59 @@ func TestHandleNexusCompletion_Canceled(t *testing.T) {
...
+			Outcome: &persistencespb.ChasmNexusCompletion_Success{
+				Success: &commonpb.Payload{Data: []byte("ok")},
+			},
+			CloseTime: timestamppb.New(time.Now()),
+		},
+		expectPaused: false,
+		expectStatus: enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED,
+		validateInvoker: func(t *testing.T, invoker *scheduler.Invoker) {
+			var deferred *schedulespb.BufferedStart
+			for _, start := range invoker.BufferedStarts {
```
Comment (davidporter-id-au) [inline chasm/lib/scheduler/scheduler_nexus_completion_test.go:294]: would we expect BufferedStart ordering to ever change?

File: tests/schedule_test.go [hunk 25]
```diff
@@ -74,66 +75,252 @@ func scheduleCommonOpts(t *testing.T) []testcore.TestOption {
...
+			Identity:          "test",
+			RequestId:         uuid.NewString(),
+		})
+		require.NoError(t, err)
+	}
+	return len(running)
+}
+
 func TestScheduleCHASM(t *testing.T) {
+	t.Parallel()
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:240]: No change required: 

Kinda surprised this works? I guess there's no shared state. Might be on the upper bound of what a typical system can do in terms of opening up ports and system resources simultaneously

File: tests/schedule_test.go [hunk 26]
```diff
@@ -3672,342 +4375,315 @@ func TestScheduleCreationRolloutPercent(t *testing.T) {
 	}
 }
 
-// testSingleDateScheduleCloses verifies that a CHASM schedule configured with
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:3670]: This is superceded by testScheduleClosesFromIdle/singleDate I suppose

File: tests/schedule_test.go [hunk 27]
```diff
@@ -202,6 +389,139 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+}
+
+// testFutureActionTimesAdvanceWhilePaused verifies that ListSchedules returns
+// up-to-date FutureActionTimes for a paused schedule. CHASM's always-on
+// Generator advances the high water mark and rebuilds FutureActionTimes
+// against the spec even when paused, so listed times stay in the future and
+// never roll into the past. The legacy V1 scheduler workflow does not advance
+// while paused, so its projected times would freeze at pause time and
+// eventually all sit in the past - hence this test is registered CHASM-only.
+func testFutureActionTimesAdvanceWhilePaused(t *testing.T, newContext contextFactory) {
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:452]: Nice: This is a great test, I learned a bunch by reading it

File: tests/schedule_test.go [hunk 28]
```diff
@@ -202,6 +389,139 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+	})
+	require.Zero(t, runs.Load(), "a paused schedule must not fire any workflows")
+}
+
+// testBufferOneDeferredFiresAfterCompletion exercises the BUFFER_ONE deferred
+// lifecycle end-to-end: an action that gets buffered while a workflow is
+// running must fire once that workflow completes. Without re-enabling the
+// deferred start (Attempt=-1 -> 0 in recordCompletedAction), the buffered
+// fire would be stranded.
+func testBufferOneDeferredFiresAfterCompletion(t *testing.T, newContext contextFactory) {
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:490]: This is really helpful for understanding the expected behaviour.

What's the desired behaviour for ScheduleOverlapPolicyBufferOne when the workflows buffered can't keep up and hit the buffer limit?

File: tests/schedule_test.go [hunk 29]
```diff
@@ -3479,107 +3803,486 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
-	schedule := &schedulepb.Schedule{
+	var runs atomic.Int32
+	registerCountingWorkflow(s, wt, &runs)
+
+	// Single calendar entry a few seconds in the future. While paused, its
+	// time will pass. Offset must exceed worst-case CreateSchedule latency
+	// under parallel load so the entry is still genuinely in the future when
+	// the server materializes the schedule - otherwise the test passes for
+	// the wrong reason (HWM already past the entry at create time).
+	fireAt := time.Now().Add(10 * time.Second).UTC()
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:3847]: Not worth changing now, but might be worth introducing synctest: https://go.dev/blog/testing-time

File: tests/schedule_test.go [hunk 30]
```diff
@@ -3479,107 +3803,486 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+		ScheduleId: sid,
+	})
+	require.NoError(t, err)
+	require.NotEmpty(t, desc.Info.RecentActions, "RecentActions should include the manual trigger")
+	require.NotEmpty(t, desc.Info.FutureActionTimes, "schedule should still have future automated actions planned")
+}
+
+// testTriggerImmediatelyOnPausedSchedule verifies that TriggerImmediately fires
+// an action even when the schedule is paused. Manual starts bypass the paused
+// gate via useScheduledAction's Manual carve-out in processBuffer.
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:4023]: Comment, no action required: 

damn, as a product choice, that's a way to increase the API surface area :/

File: tests/schedule_test.go [hunk 31]
```diff
@@ -74,66 +75,252 @@ func scheduleCommonOpts(t *testing.T) []testcore.TestOption {
...
+		DayOfMonth: fmt.Sprintf("%d", at.Day()),
+		Hour:       fmt.Sprintf("%d", at.Hour()),
+		Minute:     fmt.Sprintf("%d", at.Minute()),
+		Second:     fmt.Sprintf("%d", at.Second()),
+	}
+}
+
+// registerCountingWorkflow registers a workflow that records each execution in
+// runs (via SideEffect, so replays don't double-count) and returns immediately.
+func registerCountingWorkflow(env *testcore.TestEnv, wt string, runs *atomic.Int32) {
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:188]: Comment: No action required 

It took me quite a bit of staring to understand the intent here, I guess it's intent is to count the number of independent invocations of this particular workflow per pointer/call-site by relying on the atomic bump inside the side-effect. I guess it works, but it initially looked wrong until i realised nothing was being passed directly to the workflow, it's just mutating the register. 

I guess it might need a small warning sign to not be used more than once per scoped registration as implicitly it's kind of a (per registration) global.

File: tests/schedule_test.go [hunk 32]
```diff
@@ -3479,107 +3803,486 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+	createSchedule(ctx, t, s, sid, &schedulepb.Schedule{
+		Spec:   intervalSpec(fastInterval),
+		State:  &schedulepb.ScheduleState{Paused: true},
+		Action: startWorkflowAction(s, wid, wt),
+	})
+
+	now := time.Now().UTC()
+	patchSchedule(ctx, t, s, sid, backfillPatch(now.Add(-5*time.Second), now, enumspb.SCHEDULE_OVERLAP_POLICY_SKIP))
+
+	// SKIP collapses the 5-tick backfill to exactly one fire, and it stays there.
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:4219]: Question: Was this just matching the v1 behaviour? It just seems both extremely specific but also slightly idiosyncratic. No worries either way, but it's just not an obvious design choice I guess.

File: tests/schedule_test.go [hunk 33]
```diff
@@ -3479,107 +3803,486 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+	ctx := newContext(s.Context())
+	createSchedule(ctx, t, s, sid, schedule)
+
+	// Update with an oversized request ID.
+	_, err := s.FrontendClient().UpdateSchedule(ctx, &workflowservice.UpdateScheduleRequest{
+		Namespace:  s.Namespace().String(),
+		ScheduleId: sid,
+		Schedule:   schedule,
+		Identity:   "test",
+		RequestId:  strings.Repeat("x", 1001),
```
Comment (davidporter-id-au) [inline tests/schedule_test.go:4254]: (No action required) We don't validate these as being uuids? Sadness

**Review by davidporter-id-au — APPROVED** [review]

File: chasm/lib/scheduler/invoker_execute_task_test.go (hunk 22 above)
Comment (davidporter-id-au) [reply chasm/lib/scheduler/invoker_execute_task_test.go:562]: claude explained it to me: 
```
RunId being set means the workflow has already been started successfully. The BufferedStart is the scheduler's record of an action, and its lifecycle is roughly:

1. Pending: RunId == "", Attempt > 0 → this is what ExecuteTask acts on; it calls StartWorkflowExecution.
2. Started: StartWorkflowExecution succeeds → recordExecuteResult writes the returned RunId and StartTime back onto the entry (invoker.go:208-209).
3. Completed: a Nexus completion callback later marks Completed.
```
which I'd forgotten about. I guess weakly, maybe adding a small detail about what the runID being set implies might be worth adding to the string, but not worth holding the PR up over.

File: tests/schedule_test.go (hunk 29 above)
Comment (davidporter-id-au) [reply tests/schedule_test.go:3847]: oh, this doesn't work with RPCs as @fretz12  I guess. 🤦

File: tests/schedule_test.go (hunk 25 above)
Comment (chaptersix) [reply tests/schedule_test.go:240]: s := testcore.NewEnv(t, scheduleCommonOpts(t)...)

shares clusters and each tests is isolated in it's own namespace. 

Stephan has put a lot of time and effort optimizing this. goat

File: tests/schedule_test.go [hunk 34]
```diff
@@ -202,6 +389,139 @@ func testBufferSizeReportedWhenBuffered(t *testing.T, newContext contextFactory)
...
+		})
+		return descErr == nil && desc.GetInfo().GetBufferSize() == 1 && len(desc.GetInfo().GetRunningWorkflows()) == 1
+	}, awaitTimeout, pollInterval, "expected exactly one running workflow with one deferred start buffered behind it")
+	require.Equal(t, int32(1), runs.Load(), "only the first workflow should have fired before the running one completes")
+
+	// Releasing the running workflow must re-enable the deferred start (Attempt=-1 -> 0) so it fires.
+	require.Equal(t, 1, completeRunningWorkflows(ctx, t, s, sid))
+	await.RequireTruef(t, func() bool { return runs.Load() == 2 },
+		awaitTimeout, pollInterval,
+		"deferred start must fire after the running workflow completes - regression for the Attempt=-1 -> 0 re-enable path")
```
Comment (chaptersix) [inline tests/schedule_test.go:525]: IIUC the action in the buffer should immediately start not wait until the next scheduled action time. For skip, it will wait until the next action time. 

is there a check we can add that insures the action that is started after the completion of the previous action is the one that was buffered and not just the next action?

File: tests/schedule_test.go [hunk 35]
```diff
@@ -3479,107 +3803,486 @@ func testCHASMUnpauseResumesProcessing(t *testing.T, newContext contextFactory)
...
+	windowStart := prevHour.Add(20 * time.Minute)
+	windowEnd := windowStart.Add(10 * time.Second)
+	patchSchedule(ctx, t, s, sid, backfillPatch(windowStart, windowEnd, enumspb.SCHEDULE_OVERLAP_POLICY_ALLOW_ALL))
+
+	// No spec tick falls inside a sub-interval window, so no action fires.
+	require.Never(t, func() bool { return runs.Load() > 0 },
+		neverWindow, pollInterval,
+		"backfill range narrower than spec interval must produce no actions")
+
+	// And the schedule must still be describable - the backfiller drained without
```
Comment (chaptersix) [inline tests/schedule_test.go:4188]: nit: the test doesn't validate the scheduler's backfiller drained cleanly. I don't think it can unless we describe the mutable state.

File: tests/schedule_test.go (hunk 32 above)
Comment (chaptersix) [reply tests/schedule_test.go:4219]: Yeah, I have the same question.

**Review by chaptersix — APPROVED** [review]
LGTM aside from a few comments. Some of which are for my own education.

File: chasm/lib/scheduler/generator_tasks_test.go (hunk 1 above)
Comment (lina-temporal) [reply chasm/lib/scheduler/generator_tasks_test.go:203]: Correct on both counts, yep.

File: chasm/lib/scheduler/invoker_execute_task_test.go (hunk 22 above)
Comment (lina-temporal) [reply chasm/lib/scheduler/invoker_execute_task_test.go:562]: added a comment

File: chasm/lib/scheduler/invoker_execute_task_test.go (hunk 23 above)
Comment (lina-temporal) [reply chasm/lib/scheduler/invoker_execute_task_test.go:575]: Updated with a comment.

File: chasm/lib/scheduler/scheduler_nexus_completion_test.go (hunk 24 above)
Comment (lina-temporal) [reply chasm/lib/scheduler/scheduler_nexus_completion_test.go:294]: I think so, yeah, since each start can have a different policy, and therefore could be removed (or kept around) no matter where in the slice it happens to be.

File: tests/schedule_test.go (hunk 25 above)
Comment (lina-temporal) [reply tests/schedule_test.go:240]: +1^, the tests were just getting painfully slow, so `t.Parallel()` was sort of a must with this change

File: tests/schedule_test.go (hunk 28 above)
Comment (lina-temporal) [reply tests/schedule_test.go:490]: Just checked - right now, the limit's only applied to backfillers (which will simply backoff and retry, keeping their existing HWM, isntead of dropping starts). Generated actions will pile up (and they shouldn't, they should be limited to the buffer size), unless the schedule also has a `LimitedActions` set. Will send a PR to fix that.

File: tests/schedule_test.go (hunk 30 above)
Comment (lina-temporal) [reply tests/schedule_test.go:4023]: Yeah, and makes the invariants all the more confusing.

File: tests/schedule_test.go (hunk 31 above)
Comment (lina-temporal) [reply tests/schedule_test.go:188]: Ack, yeah, I'll add a small comment that `runs` is meant to be an `out` pointer :)

File: tests/schedule_test.go (hunk 32 above)
Comment (lina-temporal) [reply tests/schedule_test.go:4219]: Yep, this is the same across V1 and V2. I mean, going further, if you had a backfill whose policy was `SKIP` *and* the workflow had an automated action running, the user would just see.. nothing happen. IMO, today, this is probably best addressed with hints in the UI and more visible audit logging - but when we're fully on V2, we can start returning results for `Patch`, and part of that result could be something indicative of how the backfill did/would play out.

File: tests/schedule_test.go (hunk 33 above)
Comment (lina-temporal) [reply tests/schedule_test.go:4254]: correct, they can technically be anything; I think foundations *would* like to validate them as UUIDs, but at this point, cat's prob outta the bag on BYOID

File: tests/schedule_test.go (hunk 34 above)
Comment (lina-temporal) [reply tests/schedule_test.go:525]: > IIUC the action in the buffer should immediately start not wait until the next scheduled action time

Correct - that's what the tests are showing, that as soon as the running workflow blocking the buffered action completes, the buffered action fires.

> is there a check we can add that insures the action that is started after the completion of the previous action is the one that was buffered and not just the next action?

Yes, actually, we can check its nominal time. Will update the test.

File: tests/schedule_test.go (hunk 35 above)
Comment (lina-temporal) [reply tests/schedule_test.go:4188]: Yep, agree.
